### ISSUE TITLE
[Bug]: Opening storybook-static/index.html in browser does not work anymore in v7

### ISSUE THREAD

**Issue by FlorianChristie** [storybookjs/storybook#21978]
### Describe the bug

When trying to open the output generated from `storybook build` (storybook-static/index.html) in Chromium (or Chrome or Firefox) only a blank page is shown. The console errors suggest that this is a problem with the CORS policy:
`Access to script at 'file:///home/username/test-sb/storybook-static/sb-manager/runtime.mjs' from origin 'null' has been blocked by CORS policy: Cross origin requests are only supported for protocol schemes: http, data, isolated-app, chrome-extension, chrome, https, chrome-untrusted.`

When starting Chromium with disabled web security (`chromium --disable-web-security --user-data-dir="[storybook-test]"`) the storybook can be opened. 
Is there any way to get this running again without having to start Chromium in this mode? We did rely on this as we do not deploy our storybook anywhere but rather just archive the static files.

### To Reproduce

`mkdir test-sb && cd test-sb`
`npm init`
`npm install react react-dom`
`npx sb init`
`npm run build-storybook`
Open `/home/username/test-sb/storybook-static/index.html` in Chromium

### System

```shell
Environment Info:

  System:
    OS: Linux 5.15 Ubuntu 20.04.6 LTS (Focal Fossa)
    CPU: (12) x64 Intel(R) Core(TM) i7-10750H CPU @ 2.60GHz
  Binaries:
    Node: 18.15.0 - ~/.nvm/versions/node/v18.15.0/bin/node
    npm: 9.5.0 - ~/.nvm/versions/node/v18.15.0/bin/npm
  Browsers:
    Chrome: 112.0.5615.49
    Firefox: 112.0
  npmPackages:
    @storybook/addon-essentials: ^7.0.2 => 7.0.2 
    @storybook/addon-interactions: ^7.0.2 => 7.0.2 
    @storybook/addon-links: ^7.0.2 => 7.0.2 
    @storybook/blocks: ^7.0.2 => 7.0.2 
    @storybook/react: ^7.0.2 => 7.0.2 
    @storybook/react-webpack5: ^7.0.2 => 7.0.2 
    @storybook/testing-library: ^0.0.14-next.2 => 0.0.14-next.2
```

**Comment by mkosir** [comment]
Same issue happening for me, its quite blocking imo, since we can't build and deploy our storybooks.

**Comment by mert-duzgun** [comment]
Hi, v7 docs recommending using a server to see your build output. Just run `npx http-server storybook-static`. 

[Reference](https://storybook.js.org/docs/react/faq#i-see-a-no-preview-error-with-a-storybook-production-build)

**Comment by ghost91-** [comment]
@mert-duzgun using a server is not an (easy) option for us. So far, we have simply built the static stroybook in Jenkins and served it as Jenkins artifact. If we really need a full web server for this, we would need to set up quite a bit of additional infrastructure (for each storybook—yes, we have multiple...). Of course it's possible, but we really appreciate the simplicity of the previous solution.

**Comment by mkosir** [comment]
@shilman happy user of v7 here :), just wondering what is the reason for this "change" in build output, I assume storybook build process (and tooling) got much more complex, and can't be served as before in v6?

**Comment by PupoSDC** [comment]
+1 to the list. This makes our life significanently harder. 
For me the problem is that storybook is available here: 

`https://chair-flight-cxdu0ds6v-puposdc-s-team.vercel.app/storybook` 

and its trying to fetch from the wrong directory. Ie.e: 

https://www.chair-flight.com/sb-addons/essentials-backgrounds-2/manager-bundle.mjs

instead of 

https://www.chair-flight.com/storybook/sb-addons/essentials-backgrounds-2/manager-bundle.mjs

**Comment by itwaze** [comment]
@FlorianChristie the issue is that **file://** protocol doesn't support **import/export**
https://www.w3schools.com/js/js_modules.asp
![image](https://user-images.githubusercontent.com/10471508/232690657-d82db3e8-ee49-4bbf-b8d2-af7fc10adeca.png)

in order to serve the build locally, one of the way just install VSCode extension calling **LiveServer (Five Server)**
https://github.com/yandeu/five-server-vscode

then by pressing 1 button you will start to serve your opened folder
![image](https://user-images.githubusercontent.com/10471508/232691280-3cbcac2c-2861-4c54-b5c3-20ef52483375.png)

in this way you will not have any extra scripts in your **package.json**

**Comment by shilman** [comment]
My understanding, based on convo with the other maintainers, is that we are using modern ESM now and the browser CORS rules are more restrictive for browsing file  URLs with ESM.

@ndelangen did I get that right?

**Comment by PupoSDC** [comment]
> +1 to the list. This makes our life significantly harder.
> 
> For me the problem is that storybook is available here:
> 
> [https://chair-flight-cxdu0ds6v-puposdc-s-team.vercel.app/storybook`](https://chair-flight-cxdu0ds6v-puposdc-s-team.vercel.app/storybook%60)
> 
> and its trying to fetch from the wrong directory. Ie.e:
> 
> https://chair-flight-cxdu0ds6v-puposdc-s-team.vercel.app/sb-addons/essentials-backgrounds-2/manager-bundle.mjs
> 
> instead of
> 
> https://chair-flight-cxdu0ds6v-puposdc-s-team.vercel.app//storybook/sb-addons/essentials-backgrounds-2/manager-bundle.mjs

FWIW, I got around my particular issue by using next js redirect: 

```ts
module.exports = withNx(
  withBundleAnalyzer(
    withMDX({
      async rewrites() {
        return [
          "assets",
          "sb-addons",
          "sb-common-assets",
          "sb-manager",
          "sb-preview",
          "index.json",
          "iframe.html",
          "project.json",
          "stories.json",
        ].map((source) => ({
          source: `/${source}/:path*`,
          destination: `/storybook/${source}/:path*`,
        }));
      },
    })
  )
);
```

This might become a problem if we ever decide to have more than one SB... but realisticly we won't so I think we are good for now :D

**Comment by FlorianChristie** [comment]
Sorry, I made that a bit unclear in my initial message. For me personally, there is no problem with storybook  v7 for local development. But since we share the static pages with other people (e.g. UI designers) that mostly don't have node installed we relied on just archiving the static story book pages e.g. in Jenkins. This is basically the problem that @ghost91- and @PupoSDC described.
And as @ghost91- pointed out: It is possible with the new approach, no doubt, but we appreciated the possibility to simply archive the static output.

**Comment by binarykitchen** [comment]
It's a breaking change when migrating from v6, so this should be mentioned in the migrating docs.

**Comment by eifr-vault** [comment]
https://github.com/storybookjs/storybook/issues/20157#issuecomment-1348881515 <- solved it for us. we're serving from s3 with nginx

**Comment by FlorianChristie** [comment]
Yup, it's there https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#dropped-support-for-file-urls

We also find a workaround meanwhile by adapting our Jenkins CSP.

I will close this issue, since it is mentioned in the official migration for v7 and thus the new wanted behavior. :slightly_smiling_face: 

Thanks for the discussion and all the hints and comments. :heart:

**Comment by sid2u** [comment]
FWIW, it's relatively easier to work around this in Firefox. It has a `security.fileuri.strict_origin_policy` setting you can disable:
![image](https://github.com/storybookjs/storybook/assets/8316961/af147fc6-4530-43a4-8001-5160657ffa77)

Might be sufficient for some use cases.

**Comment by mellis481** [comment]
>> @shilman browsers not supporting ESM natively when on the file://-protocol is an missed oversight of the modernization work, unfortunate.

> I'm guessing the issue can be closed with "won't fix" if it's purely about the fact this protocol doesn't work anymore.
It certainly wasn't our intention to break that feature, though how easy it is nowadays to simply spawn a quick http server that works great, it also doesn't feel super useful anymore.

@ndelangen @FlorianChristie I just want to share our use-case for a static storybook build running from the file system because I'm interpreting the response above as "oh well."

We were using static storybook builds previous to v7 to share in-progress design system work with designers and PMs. Being able to build storybook, zip up the static assets, and instruct non-developers to just open the unzipped `index.html` was extremely convenient. Now, these same folks would need to do the following which I don't think is reasonable for non-developers:
- install Node.js
- install npx globally
- run `npx http-server storybook-static`

**Comment by shilman** [comment]
@mellis481 I don't know if there's any easy fix for this -- perhaps @ndelangen knows. FWIW, you can share Storybooks on https://www.chromatic.com/ for free, including access control. I know there are scenarios where sending files around might be preferred, but perhaps this is a suitable workaround.
